WowApiMockWrapper.lua

```diff
@@ -0,0 +1,122 @@
+    MockMethodPartial(wrapper, "UnitRace", "Gnome")
+    MockMethodPartial(wrapper, "UnitFactionGroup", "Alliance")
+    MockMethodPartial(wrapper, "IsInRaid", true)
+    MockMethodPartial(wrapper, "UnitClass", {"Mage", "MAGE", 8})
+    MockMethodPartial(wrapper, "UnitExists", 1)
+    MockMethodPartial(wrapper, "UnitGUID", "0xF530007EAC083004")
+    MockMethodPartial(wrapper, "UnitAffectingCombat", 1)
+    MockMethodPartial(wrapper, "GetSpecialization", 1234)
+    MockMethodPartial(wrapper, "GetActiveSpecGroup", 1234)
+    MockMethodPartial(wrapper, "GetInstanceInfo", {"The Emerald Nightmare", "raid", 10, "10-30 Player Heroic", 10, 0, true, 1234, 11})
+    MockMethodPartial(wrapper, "IsInInstance", {true, "raid"})
+    MockMethodPartial(wrapper, "GetNumGroupMembers", 11)
+    MockMethodPartial(wrapper, "UnitIsUnit", true)
+    MockMethodPartial(wrapper, "GetRaidRosterInfo", {"Guildmate", 2, 1, 110, "Mage", "MAGE", "The Emerald Nightmare", 1, nil, "MAINTANK", 1})
+    MockMethodPartial(wrapper, "GetSpecialization", 1)
+    MockMethodPartial(wrapper, "GetSpecializationRole", "TODO: Mock GetSpecializationRole")
+    MockMethodPartial(wrapper, "UnitInVehicle", nil)
+    MockMethodPartial(wrapper, "UnitHasVehicleUI", false)
+    MockMethodPartial(wrapper, "GetSpellInfo", {"Fireball", "Rank 5", "SOME_ICON_PATH", 1500, 0, 40, 1234})
+    MockMethodPartial(wrapper, "SendChatMessage", nil)
+    MockMethodPartial(wrapper, "GetChannelName", {1234, "The Emerald Nightmare - General", 0})
+    MockMethodPartial(wrapper, "UnitInBattleground", 13)
+    MockMethodPartial(wrapper, "UnitInRaid", 1)
+    MockMethodPartial(wrapper, "UnitInParty", 1)
+    MockMethodPartial(wrapper, "PlaySoundFile", nil)
+    --MockMethodPartial(wrapper, "PlaySoundKitID", "TODO: Mock PlaySoundKitID")
+    MockMethodPartial(wrapper, "GetTime", 123456789)
+    --MockMethodPartial(wrapper, "GetSpellLink", "TODO: Mock GetSpellLink")
+    --MockMethodPartial(wrapper, "GetItemInfo", "TODO: Mock GetItemInfo")
+    --MockMethodPartial(wrapper, "CreateFrame", "TODO: Mock CreateFrame")
+    MockMethodPartial(wrapper, "IsShiftKeyDown", nil)
+    MockMethodPartial(wrapper, "GetScreenWidth", 1024)
+    MockMethodPartial(wrapper, "GetScreenHeight", 768)
+    MockMethodPartial(wrapper, "GetCursorPosition", {600, 600})
+    --MockMethodPartial(wrapper, "random", "TODO: Mock random")
+    MockMethodPartial(wrapper, "UpdateAddOnCPUUsage", nil)
+    MockMethodPartial(wrapper, "GetFrameCPUUsage", {100, 20})
+    MockMethodPartial(wrapper, "debugprofilestop", 100)
+    MockMethodPartial(wrapper, "debugstack", "FAKE_DEBUG_STACK")
+    MockMethodPartial(wrapper, "IsSpellKnown", true)
+    MockMethodPartial(wrapper, "GetAddOnEnableState", 2)
+    --MockMethodPartial(wrapper, "IsPlayerMoving", "TODO: Mock IsPlayerMoving")
+    MockMethodPartial(wrapper, "UnitIsDead", nil)
+end
+
+-- TESTING When we change the mock on one instance it doesn't affect others.
```

Comment:
There are several tests at the end of this file right now to make sure that as I'm modifying the way the mock works I don't break any of the important functionality.

Things tested here are:

- Each mock can be modified without affecting other mocks.
- When the mocked data is a function we call the function with all the right parameters and return the correct values.
- When the mock has multiple return values we make sure they come back the way they're supposed to.
- When the mock returns "nil" we don't lose it along the way.
- When we assign one of the mocked functions into a local variable and later call it using the local it still works.